src/detect-tls-validity.c

```diff
@@ -0,0 +1,1008 @@
+    parse_regex_study = pcre_study(parse_regex, 0, &eb);
+    if (eb != NULL) {
+        SCLogError(SC_ERR_PCRE_STUDY,"pcre study failed: %s", eb);
+        goto error;
+    }
+
+error:
+    return;
+}
+
+/**
+ * \internal
+ * \brief Function to match validity field in a tls certificate.
+ *
+ * \param t       Pointer to thread vars.
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param s       Pointer to the Signature.
+ * \param m       Pointer to the sigmatch that we will cast into
+ *                DetectTlsValidityData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectTlsValidityMatch (ThreadVars *t, DetectEngineThreadCtx *det_ctx,
+                   Flow *f, uint8_t flags, void *state, Signature *s, SigMatch *m)
+{
+    SCEnter();
+
+    SSLState *ssl_state = (SSLState *)state;
+    if (ssl_state == NULL) {
+        SCLogDebug("no tls state, no match");
+        SCReturnInt(0);
+    }
+
+    int ret = 0;
+
+    SSLStateConnp *connp = NULL;
+    if (flags & STREAM_TOSERVER)
+        connp = &ssl_state->client_connp;
+    else
+        connp = &ssl_state->server_connp;
+
+    const DetectTlsValidityData *dd = (const DetectTlsValidityData *)m->ctx;
+
+    time_t cert_epoch = 0;
+    if (dd->type == DETECT_TLS_TYPE_NOTBEFORE)
+        cert_epoch = connp->cert0_not_before;
+    else if (dd->type == DETECT_TLS_TYPE_NOTAFTER)
+        cert_epoch = connp->cert0_not_after;
+
+    if (cert_epoch == 0)
+        SCReturnInt(0);
+
+    if (dd->mode == DETECT_TLS_VALIDITY_EQ && dd->epoch == cert_epoch)
+        ret = 1;
+    else if (dd->mode == DETECT_TLS_VALIDITY_LT && dd->epoch > cert_epoch)
+        ret = 1;
+    else if (dd->mode == DETECT_TLS_VALIDITY_GT && dd->epoch < cert_epoch)
+        ret = 1;
+    else if (dd->mode == DETECT_TLS_VALIDITY_RA &&
+            dd->epoch < cert_epoch && dd->epoch2 > cert_epoch)
+        ret = 1;
+
+    SCReturnInt(ret);
+}
+
+/**
+ * \internal
+ * \brief Function to convert date string to epoch.
+ *
+ * \param string Date string.
+ *
+ * \retval epoch on success.
+ * \retval 0 on failure.
+ */
+static time_t DateStringToEpoch (char *string)
+{
+    int r = 0;
+    struct tm tm;
+    char *patterns[] = {
+            /* ISO 8601 */
+            "%Y-%m",
+            "%Y-%m-%d",
+            "%Y-%m-%d %H",
+            "%Y-%m-%d %H:%M",
+            "%Y-%m-%d %H:%M:%S",
+            "%Y-%m-%dT%H",
+            "%Y-%m-%dT%H:%M",
+            "%Y-%m-%dT%H:%M:%S",
+            "%H:%M",
+            "%H:%M:%S",
+            "%Y-%j",
+            /* epoch */
+            "%s"
+    };
+
+    r = SCStringPatternToTime(string, patterns, 12, &tm);
+
+    if (r != 0)
+        return 0;
+
+    time_t ret;
+    int gmtoff;
+    char *tz;
+
+    gmtoff = tm.tm_gmtoff;
+
+    /* from timegm */
+    tz = getenv("TZ");
+    if (tz)
+        tz = strdup(tz);
+    setenv("TZ", "", 1);
```

Comment:
Note to self.

Replacing line 195 - 216 with:
```c
return mktime(&tm) + tm.tm_gmtoff;
```
looks like it might do the same without touching any environment variables...